File: linkgame-client/src/main/java/io/github/octcarp/linkgame/client/net/ServerHandlerThread.java
package io.github.octcarp.linkgame.client.net;

import io.github.octcarp.linkgame.common.model.Game;
import io.github.octcarp.linkgame.common.model.Match;
import io.github.octcarp.linkgame.common.packet.Response;
import io.github.octcarp.linkgame.common.packet.SimpStatus;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.net.Socket;
import java.util.List;

public class ServerHandlerThread implements Runnable {
    private final Socket socket;
    private final ObjectInputStream ois;

    public ServerHandlerThread(Socket socket, ObjectInputStream ois) {
        this.socket = socket;
        try {
            this.ois = ois;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void run() {
        try {
            while (true) {
                Object oin = ois.readObject();
                if (!(oin instanceof Response)) {
                    continue;
                }
                Response response = (Response) oin;
                switch (response.getType()) {
                    case LOGIN_RESULT -> {
                        SimpStatus status = (SimpStatus) response.getData();
                        LoginData.getInstance().rePlayerLogin(status);
                    }
                    case REGISTER_RESULT -> {
                        SimpStatus status = (SimpStatus) response.getData();
                        LoginData.getInstance().rePlayerRegister(status);
                    }
                    case LOGOUT_RESULT -> {
                        SimpStatus status = (SimpStatus) response.getData();
                        LoginData.getInstance().reLogout(status);
                    }
                    case ALL_WAITING_PLAYERS -> {
                        List<String> players = (List<String>) response.getData();
                        LobbyData.getInstance().reAllWaitingPlayers(players);
                    }
                    case START_MATCH -> {
                        Match match = (Match) response.getData();
                        MatchData.getInstance().reStartMatch(match);
                    }
                    case SYNC_MATCH -> {
                        Match match = (Match) response.getData();
                        MatchData.getInstance().reSyncMatch(match);
                    }
                    case SYNC_BOARD -> {
                        Game game = (Game) response.getData();
                        MatchData.getInstance().reSyncBoard(game);
                    }
                    case MATCH_FINISHED -> {
                        Match match = (Match) response.getData();
                        MatchData.getInstance().reMatchFinished(match);
                    }
                }
            }
        } catch (IOException e) {
//            SceneSwitcher.getInstance().netErrAndReturn();
            throw new RuntimeException(e);
        } catch (ClassNotFoundException | ClassCastException e) {
            throw new RuntimeException(e);
        } finally {
            ClientService.getInstance().disconnect();
        }
    }
}

File: game-client/src/main/kotlin/io/github/octcarp/sustech/cs209a/linkgame/client/net/ServerHandlerThread.kt
package io.github.octcarp.sustech.cs209a.linkgame.client.net

import io.github.octcarp.sustech.cs209a.linkgame.client.utils.SceneSwitcher
import io.github.octcarp.sustech.cs209a.linkgame.common.model.Game
import io.github.octcarp.sustech.cs209a.linkgame.common.model.Match
import io.github.octcarp.sustech.cs209a.linkgame.common.model.MatchRecord
import io.github.octcarp.sustech.cs209a.linkgame.common.packet.Response
import io.github.octcarp.sustech.cs209a.linkgame.common.packet.ResponseType
import io.github.octcarp.sustech.cs209a.linkgame.common.packet.SimpStatus
import java.io.IOException
import java.io.ObjectInputStream
import java.net.Socket

class ServerHandlerThread(private val socket: Socket, private val ois: ObjectInputStream) : Runnable {
    override fun run() {
        try {
            while (true) {
                val oin = ois.readObject()
                if (oin !is Response) {
                    continue
                }
                val response = oin
                when (response.type) {
                    ResponseType.LOGIN_RESULT -> {
                        val status = response.data as SimpStatus
                        LoginData.rePlayerLogin(status)
                    }

                    ResponseType.REGISTER_RESULT -> {
                        val status = response.data as SimpStatus
                        LoginData.rePlayerRegister(status)
                    }

                    ResponseType.LOGOUT_RESULT -> {
                        val status = response.data as SimpStatus
                        LoginData.reLogout(status)
                    }

                    ResponseType.GET_MATCH_RECORD_RESULT -> {
                        val matches = (response.data as List<MatchRecord>)
                        RecordData.reSyncRecord(matches.toMutableList())
                    }

                    ResponseType.ALL_WAITING_PLAYERS -> {
                        val players = (response.data as List<String>)
                        LobbyData.reAllWaitingPlayers(players.toMutableList())
                    }

                    ResponseType.START_MATCH -> {
                        val match = response.data as Match
                        MatchData.reStartMatch(match)
                    }

                    ResponseType.SYNC_MATCH -> {
                        val match = response.data as Match
                        MatchData.reSyncMatch(match)
                    }

                    ResponseType.SYNC_BOARD -> {
                        val game = response.data as Game
                        MatchData.reSyncBoard(game)
                    }

                    ResponseType.RECONNECT_SUCCESS -> {
                        val match = response.data as Match
                        MatchData.reConnectToMatch(match)
                    }

                    ResponseType.MATCH_FINISHED -> {
                        val match = response.data as Match
                        MatchData.reMatchFinished(match)
                    }

                    ResponseType.NO_MATCH_TO_RECONNECT -> {
                        LobbyData.reNoReconnect()
                    }

                    ResponseType.WAITING_OPPONENT -> TODO()
                    ResponseType.OPP_DISCONNECTED -> TODO()
                    ResponseType.OPP_RECONNECTED -> TODO()
                    ResponseType.ERROR_MESSAGE -> TODO()
                }
            }
        } catch (e: IOException) {
            SceneSwitcher.netErrAndReturn();
            throw RuntimeException(e)
        } catch (e: ClassNotFoundException) {
            throw RuntimeException(e)
        } catch (e: ClassCastException) {
            throw RuntimeException(e)
        } finally {
            ClientService.disconnect()
        }
    }
}
